### DISCUSSION TITLE
Ways to hide subcategory in iModel viewer

### DISCUSSION THREAD

**Discussion by hsanand61564** [iTwin/itwinjs-core#948] [Q&A] [answered]
Hi

We are displaying imodel file into the imodel viewer. But when it comes to customization of subcategory thing it is not working for me. I want to make the subcategory invisible to true. 

Following is the code snippet that I am using

1.
![image](https://user-images.githubusercontent.com/60655118/110941666-bb736b00-835e-11eb-8202-378719429c14.png)

2.
![image](https://user-images.githubusercontent.com/60655118/110942178-73a11380-835f-11eb-9eb1-674883cba3e7.png)

3.
![image](https://user-images.githubusercontent.com/60655118/110941899-042b2400-835f-11eb-9e51-c66ec17699d4.png)

After these steps listed above still, I am unable to hide the subcategories in my model.

Can you please suggest some other way round or modifications in the steps so as to achieve the hiding of subcategory thing?

Any help would be really appreciable

**Comment by pmconne** [comment]
You have `{ invisible: false }`. That means you want the subcategory to be visible. Did you mean to use `{ invisible: true }`?

**Reply by hsanand61564** [reply]
@pmconne 
 yeah, I am using {invisible: true} only, it was just experimenting and put up the wrong screenshot.

Also after applying the I tried getting up the ViewStateProps using vp.view.toProps(); wherein **vp** is my viewport and then I used console.log to see the props. Following is the screenshot

![image](https://user-images.githubusercontent.com/60655118/110943204-f24a8080-8360-11eb-94a6-57ce9db45d49.png)

in the console as well it is showing as invisible: true for all subcategories but still not hiding them in the viewer

**Reply by pmconne** [reply]
What version of imodeljs-frontend are you using? Are you using ViewportComponent? If you change `onViewOpen.addOnce` to `onViewOpen.addListener` and stick a breakpoint inside, does it get hit more than once? Did you consider not making listViewerSettings `async`, since it does nothing asynchronous?

**Reply by hsanand61564** [reply]
@pmconne We are using 2.13.0 version of imodeljs-frontend. Yes we are using viewportComponent. 

If we change onViewOpen.addOnce to onViewOpen.addListener we get hit more than once.

yes we would be removing async from method listViewerSettings

**Reply by pmconne** [reply]
Thanks. If you are only using one ViewportComponent but your onViewOpen listener gets hit more than once, then it sounds like ViewportComponent is creating more than one viewport and you are modifying a different one than is ultimately being displayed.
If you keep `addListener` instead of `addOnce` and allow `listViewerSettings` to adjust the subcategories for every viewport, do your subcategories become hidden?

**Reply by hsanand61564** [reply]
@pmconne I tried again with addListener with a debugger inside it and it stuck only once on the debugger. And with this the subcategories are still not hidden.

below is the ViewportComponent snippet

![image](https://user-images.githubusercontent.com/60655118/110957361-8c1a2980-8371-11eb-8e7c-4afcca243f1b.png)

**Reply by hsanand61564** [reply]
@pmconne Any further views on the same as still stuck on it?

**Comment by hsanand61564** [comment]
@pmconne Hi
We tried applying color to a subcategory specifically by giving the ID and it seems to be working.
![image](https://user-images.githubusercontent.com/60655118/111157291-b9a6e300-85bc-11eb-93b8-8710f7fe9402.png)

But when we add invisible: true or transp:99 to the JSON object it does not works and it removes the color as well, as if it is ignoring the request.

![image](https://user-images.githubusercontent.com/60655118/111157557-05598c80-85bd-11eb-9141-71761090f9dc.png)


Any solution to the same would be really appreciable.

Note: we are using 2.13.0 version of imodeljs-frontend.

**Reply by pmconne** [reply]
I have thus far been unable to reproduce the problem - see #961.

**Comment by pmconne** [comment, marked as answer]
I modified [this sample](https://github.com/imodeljs/frontend-sample-showcase/blob/bc2e2025f47bc0af4b99e412d32267ce1d324a4c/src/frontend-samples/hypermodeling-sample/HyperModelingUI.tsx#L42) which also uses ViewportComponent to try to reproduce your issue. In `onViewOpen` callback I query for all subcategories and make half of them invisible. This has the expected effect: geometry on those subcategories is not displayed in the viewport.

```
    IModelApp.viewManager.onViewOpen.addOnce(async (viewport: ScreenViewport) => {
      this.setState({ viewport });
      await HyperModelingApp.enableHyperModeling(viewport);
      HyperModelingApp.markerHandler.onActiveMarkerChanged.addListener((activeMarker) => this.setState({ activeMarker }));
      await HyperModelingApp.activateMarkerByName(viewport, "Section-Left");

      // Turn half of the subcategories invisible.
      let i = 0;
      for await (const row of viewport.iModel.query("select ecinstanceid as subcat from bis.subcategory")) {
        if (i % 2 === 0)
          viewport.displayStyle.overrideSubCategory(row.subcat, SubCategoryOverride.fromJSON({ invisible: true }));

        i++;
      }
    });
```

**Reply by hsanand61564** [reply]
@pmconne It's quite strange as I tried the same piece of code that you have written to my code, but it doesn't seems to be working at my end.

**Reply by pmconne** [reply]
Sorry @hsanand61564, that sounds frustrating. I suppose you can't share the rest of your code? If not, can you compare it to the code in the sample showcase I linked to see what might differ? Or, if you based your app on one of the iModel.js samples, can you tell us which one?
